### PR TITLE
Refactor IPAM sweepers & remove SLR precheck

### PR DESCRIPTION
**Pull request by drewmullen**
### Description

3 Primary changes:

1. IPAM introduced the --cascade argument that allows you to delete ipam resources in summary from the root ipam. this will make the delete process simpler and remove the requirement to wait for allocations to be removed by the service (can take over 30m)
1. IPAM also removed the requirement to create SLR prior to deploying IPAM, instead its made when the `create-ipam` call is made.
1. had to update `ipam_pool.go` because _tags test was failing prior to changes

### Relations

Closes #0000

### Output from Acceptance Testing

```
$ make testacc TESTS=TestAccIPAM PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccIPAM'  -timeout 180m
=== RUN   TestAccIPAM_byoipIPv6
    ipam_byoip_test.go:21: Environment variable IPAM_BYOIP_IPV6_MESSAGE, IPAM_BYOIP_IPV6_SIGNATURE, or IPAM_BYOIP_IPV6_PROVISIONED_CIDR is not set
--- SKIP: TestAccIPAM_byoipIPv6 (0.00s)
=== RUN   TestAccIPAMOrganizationAdminAccount_basic
    acctest.go:644: skipping test because at least one environment variable of [AWS_ALTERNATE_PROFILE AWS_ALTERNATE_ACCESS_KEY_ID] must be set. Usage: credentials for running acceptance testing in alternate AWS account.
--- SKIP: TestAccIPAMOrganizationAdminAccount_basic (0.41s)
=== RUN   TestAccIPAMPoolAllocation_ipv4Basic
=== PAUSE TestAccIPAMPoolAllocation_ipv4Basic
=== RUN   TestAccIPAMPoolAllocation_ipv4BasicNetmask
=== PAUSE TestAccIPAMPoolAllocation_ipv4BasicNetmask
=== RUN   TestAccIPAMPoolAllocation_ipv4DisallowedCIDR
=== PAUSE TestAccIPAMPoolAllocation_ipv4DisallowedCIDR
=== RUN   TestAccIPAMPoolAllocation_multiple
=== PAUSE TestAccIPAMPoolAllocation_multiple
=== RUN   TestAccIPAMPoolCIDR_basic
=== PAUSE TestAccIPAMPoolCIDR_basic
=== RUN   TestAccIPAMPoolCIDRsDataSource_basic
=== PAUSE TestAccIPAMPoolCIDRsDataSource_basic
=== RUN   TestAccIPAMPoolDataSource_basic
=== PAUSE TestAccIPAMPoolDataSource_basic
=== RUN   TestAccIPAMPool_basic
=== PAUSE TestAccIPAMPool_basic
=== RUN   TestAccIPAMPool_tags
=== PAUSE TestAccIPAMPool_tags
=== RUN   TestAccIPAMPool_ipv6Basic
=== PAUSE TestAccIPAMPool_ipv6Basic
=== RUN   TestAccIPAMPreviewNextCIDRDataSource_ipv4Basic
=== PAUSE TestAccIPAMPreviewNextCIDRDataSource_ipv4Basic
=== RUN   TestAccIPAMPreviewNextCIDRDataSource_ipv4Allocated
=== PAUSE TestAccIPAMPreviewNextCIDRDataSource_ipv4Allocated
=== RUN   TestAccIPAMPreviewNextCIDRDataSource_ipv4DisallowedCIDR
=== PAUSE TestAccIPAMPreviewNextCIDRDataSource_ipv4DisallowedCIDR
=== RUN   TestAccIPAMPreviewNextCIDR_ipv4Basic
=== PAUSE TestAccIPAMPreviewNextCIDR_ipv4Basic
=== RUN   TestAccIPAMPreviewNextCIDR_ipv4Allocated
=== PAUSE TestAccIPAMPreviewNextCIDR_ipv4Allocated
=== RUN   TestAccIPAMPreviewNextCIDR_ipv4DisallowedCIDR
=== PAUSE TestAccIPAMPreviewNextCIDR_ipv4DisallowedCIDR
=== RUN   TestAccIPAMScope_basic
=== PAUSE TestAccIPAMScope_basic
=== RUN   TestAccIPAMScope_tags
=== PAUSE TestAccIPAMScope_tags
=== RUN   TestAccIPAM_basic
=== PAUSE TestAccIPAM_basic
=== RUN   TestAccIPAM_modify
=== PAUSE TestAccIPAM_modify
=== RUN   TestAccIPAM_cascade
=== PAUSE TestAccIPAM_cascade
=== RUN   TestAccIPAM_tags
=== PAUSE TestAccIPAM_tags
=== CONT  TestAccIPAMPoolAllocation_ipv4Basic
=== CONT  TestAccIPAMPreviewNextCIDRDataSource_ipv4Allocated
=== CONT  TestAccIPAMPoolDataSource_basic
=== CONT  TestAccIPAMScope_tags
=== CONT  TestAccIPAMPreviewNextCIDR_ipv4Allocated
=== CONT  TestAccIPAMScope_basic
=== CONT  TestAccIPAMPreviewNextCIDR_ipv4Basic
=== CONT  TestAccIPAMPreviewNextCIDR_ipv4DisallowedCIDR
=== CONT  TestAccIPAMPreviewNextCIDRDataSource_ipv4DisallowedCIDR
=== CONT  TestAccIPAM_tags
=== CONT  TestAccIPAMPool_ipv6Basic
=== CONT  TestAccIPAMPoolCIDRsDataSource_basic
=== CONT  TestAccIPAMPool_tags
=== CONT  TestAccIPAMPoolAllocation_multiple
=== CONT  TestAccIPAM_cascade
=== CONT  TestAccIPAMPool_basic
=== CONT  TestAccIPAM_modify
=== CONT  TestAccIPAMPreviewNextCIDRDataSource_ipv4Basic
=== CONT  TestAccIPAM_basic
=== CONT  TestAccIPAMPoolAllocation_ipv4DisallowedCIDR
--- PASS: TestAccIPAMPoolDataSource_basic (130.09s)
=== CONT  TestAccIPAMPoolCIDR_basic
--- PASS: TestAccIPAM_cascade (138.66s)
=== CONT  TestAccIPAMPoolAllocation_ipv4BasicNetmask
--- PASS: TestAccIPAM_basic (144.41s)
--- PASS: TestAccIPAMPool_ipv6Basic (165.63s)
--- PASS: TestAccIPAMPreviewNextCIDR_ipv4DisallowedCIDR (179.31s)
--- PASS: TestAccIPAMPreviewNextCIDRDataSource_ipv4Basic (179.33s)
--- PASS: TestAccIPAMPreviewNextCIDRDataSource_ipv4DisallowedCIDR (189.37s)
--- PASS: TestAccIPAMPoolAllocation_ipv4DisallowedCIDR (189.38s)
--- PASS: TestAccIPAMPreviewNextCIDR_ipv4Basic (190.02s)
--- PASS: TestAccIPAMPoolAllocation_ipv4Basic (219.93s)
--- PASS: TestAccIPAMScope_basic (227.50s)
--- PASS: TestAccIPAMPoolAllocation_multiple (229.11s)
--- PASS: TestAccIPAMPool_basic (236.60s)
--- PASS: TestAccIPAMPreviewNextCIDR_ipv4Allocated (238.79s)
--- PASS: TestAccIPAM_tags (248.03s)
--- PASS: TestAccIPAMPreviewNextCIDRDataSource_ipv4Allocated (250.65s)
--- PASS: TestAccIPAMScope_tags (266.07s)
--- PASS: TestAccIPAMPoolCIDRsDataSource_basic (280.27s)
--- PASS: TestAccIPAMPoolAllocation_ipv4BasicNetmask (143.08s)
--- PASS: TestAccIPAMPoolCIDR_basic (155.11s)
--- PASS: TestAccIPAM_modify (288.15s)
--- PASS: TestAccIPAMPool_tags (289.26s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ec2        293.558s
```

### Sweeper tests

I also ran tests locally to validate the new sweepers work:

run any ipam sweeper to delete the stack below:

example: `make sweep SWEEPARGS=-sweep-run=aws_vpc_ipam_pool` youll see ipam as `delete-in-progress` then shortly deleted everything

build:

```
module "basic" {
  source  = "aws-ia/ipam/aws"
#   source = "../.."

  top_cidr = ["10.0.0.0/8"]
  top_name = "basic ipam"

  pool_configurations = {
    corporate-us-west-2 = {
      description = "2nd level, locale us-west-2 pool"
      cidr        = ["10.0.0.0/16", "10.1.0.0/16"]

      sub_pools = {

        sandbox = {
          name                 = "mysandbox"
          cidr                 = ["10.0.0.0/20"]
        #   ram_share_principals = var.sandbox_ou_arn
          allocation_resource_tags = {
            env = "sandbox"
          }
        }
        dev = {
          cidr = ["10.1.0.0/20"]

          sub_pools = {
            team_a = {
              cidr                 = ["10.1.0.0/24"]
            #   ram_share_principals = var.prod_account # prod account
              locale               = "us-west-2"
            }

            team_b = {
              cidr                 = ["10.1.1.0/24"]
            #   ram_share_principals = var.prod_account # prod account
            }
          }
        }
        prod = {
          cidr   = ["10.1.16.0/20"]
          locale = "us-west-2"

          sub_pools = {
            team_a = {
              cidr                 = ["10.1.16.0/24"]
            #   ram_share_principals = var.prod_account # prod account
            }

            team_b = {
              cidr                 = ["10.1.17.0/24"]
            #   ram_share_principals = var.prod_account # prod account
            }
          }
        }
      }
    }
    us-east-1 = {
      cidr   = ["10.2.0.0/16"]
      locale = "us-east-1"

      sub_pools = {

        team_a = {
          cidr                 = ["10.2.0.0/20"]
        #   ram_share_principals = var.prod_ou_arn
        }

        team_b = {
          cidr                 = ["10.2.16.0/20"]
        #   ram_share_principals = var.prod_ou_arn
        }
      }
    }
  }
}

resource "aws_vpc" "name" {
  ipv4_ipam_pool_id = module.basic.pools_level_2["us-east-1/team_b"].id
  ipv4_netmask_length = 28
}

```

### CODE REVIEW

**Comment by AdamTylerLynch** [comment]
@ewbankkit with this refactor, the assumption is that Sweeping will always sweep all objects in the hierarchy via cascade (IPAM->POOL->CIDR). This is in contrast to how it works today, where the sweepers sweep the hierarchy bottom up (CIRD->POOL->IPAM).

The benefit is a much faster sweep, 5.8 minutes versus minimum 45 minutes. This reduced Hashi account spend, and we believe fits the normal use-case of sweeping which is from CI/CD, not a user initiated action on a dependent resource.

Thoughts?

File: internal/service/ec2/sweep.go [hunk 1]
```diff
@@ -355,29 +355,28 @@ func init() {
 
 	resource.AddTestSweepers("aws_vpc_ipam_pool_cidr_allocation", &resource.Sweeper{
```
[the remark this answers is not in the document]
Comment (drewmullen) [reply internal/service/ec2/sweep.go]: ✅ removed the other sweeper definitions

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccIPAMPool_' PKG=ec2 ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 2  -run=TestAccIPAMPool_ -timeout 180m
=== RUN   TestAccIPAMPool_basic
=== PAUSE TestAccIPAMPool_basic
=== RUN   TestAccIPAMPool_tags
=== PAUSE TestAccIPAMPool_tags
=== RUN   TestAccIPAMPool_ipv6Basic
=== PAUSE TestAccIPAMPool_ipv6Basic
=== CONT  TestAccIPAMPool_basic
=== CONT  TestAccIPAMPool_ipv6Basic
--- PASS: TestAccIPAMPool_ipv6Basic (52.34s)
=== CONT  TestAccIPAMPool_tags
--- PASS: TestAccIPAMPool_basic (82.40s)
--- PASS: TestAccIPAMPool_tags (91.96s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	148.666s
% make sweep SWEEPARGS=-sweep-run=aws_vpc_ipam
# make sweep SWEEPARGS=-sweep-run=aws_example_thing
# set SWEEPARGS=-sweep-allow-failures to continue after first failure
WARNING: This will destroy infrastructure. Use only in development accounts.
go test ./internal/sweep -v -tags=sweep -sweep=us-west-2,us-east-1,us-east-2 -sweep-run=aws_vpc_ipam -timeout 60m
2022/10/13 08:42:28 [DEBUG] Running Sweepers for region (us-west-2):
2022/10/13 08:42:28 [DEBUG] Running Sweeper (aws_vpc_ipam) in region (us-west-2)
2022/10/13 08:42:28 [INFO] Retrieved credentials from "EnvConfigCredentials"
2022/10/13 08:42:28 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2022/10/13 08:42:29 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2022/10/13 08:42:29 [DEBUG] Completed Sweeper (aws_vpc_ipam) in region (us-west-2) in 1.058384662s
2022/10/13 08:42:29 Completed Sweepers for region (us-west-2) in 1.058608618s
2022/10/13 08:42:29 Sweeper Tests for region (us-west-2) ran successfully:
	- aws_vpc_ipam
2022/10/13 08:42:29 [DEBUG] Running Sweepers for region (us-east-1):
2022/10/13 08:42:29 [DEBUG] Running Sweeper (aws_vpc_ipam) in region (us-east-1)
2022/10/13 08:42:29 [INFO] Retrieved credentials from "EnvConfigCredentials"
2022/10/13 08:42:29 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2022/10/13 08:42:29 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2022/10/13 08:42:31 [DEBUG] Completed Sweeper (aws_vpc_ipam) in region (us-east-1) in 1.430475182s
2022/10/13 08:42:31 Completed Sweepers for region (us-east-1) in 1.430496731s
2022/10/13 08:42:31 Sweeper Tests for region (us-east-1) ran successfully:
	- aws_vpc_ipam
2022/10/13 08:42:31 [DEBUG] Running Sweepers for region (us-east-2):
2022/10/13 08:42:31 [DEBUG] Running Sweeper (aws_vpc_ipam) in region (us-east-2)
2022/10/13 08:42:31 [INFO] Retrieved credentials from "EnvConfigCredentials"
2022/10/13 08:42:31 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2022/10/13 08:42:31 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2022/10/13 08:42:31 [DEBUG] Completed Sweeper (aws_vpc_ipam) in region (us-east-2) in 621.62114ms
2022/10/13 08:42:31 Completed Sweepers for region (us-east-2) in 621.643922ms
2022/10/13 08:42:31 Sweeper Tests for region (us-east-2) ran successfully:
	- aws_vpc_ipam
ok  	github.com/hashicorp/terraform-provider-aws/internal/sweep	8.416s
```

**Comment by ewbankkit** [comment]
@drewmullen Thanks for the contribution :tada: :clap:.

**Pull request merged**